shell/platform/linux/public/flutter_linux/fl_engine.h

```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
  * #FlEngine is an object that contains a running Flutter engine.
  */
 
+/**
+ * FlEnginePlatformMessageHandle:
+ *
+ * A handle used to respond to platform messages.
+ */
+typedef struct _FlEnginePlatformMessageHandle FlEnginePlatformMessageHandle;
+
+/**
+ * FlEnginePlatformMessageCallback:
+ * @engine: a #FlEngine
+ * @channel: channel message received on.
+ * @message: message content received from Dart
+ * @handle: a handle to respond to the message with
+ * @user_data: (closure): data provided when registering this callback
+ *
+ * Function called when platform messages are received.
+ */
+typedef void (*FlEnginePlatformMessageCallback)(
+    FlEngine* engine,
+    const gchar* channel,
+    GBytes* message,
+    FlEnginePlatformMessageHandle* handle,
+    gpointer user_data);
+
+/**
+ * fl_engine_set_platform_message_handler:
+ * @engine: a #FlEngine
+ * @channel: channel to listen on
+ * @callback: function to call when a message is received on this channel
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Register a callback to handle platform messages. Call
+ * fl_engine_send_platform_message_response() when this message should be
+ * responded to.
+ */
+void fl_engine_set_platform_message_handler(
+    FlEngine* engine,
+    const gchar* channel,
+    FlEnginePlatformMessageCallback callback,
+    gpointer user_data);
+
+/**
+ * fl_engine_send_platform_message_response:
+ * @engine: a #FlEngine
+ * @handle: message being responded to
+ * @response: (allow-none): response to send or %NULL for an empty response.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Respond to a platform message.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_engine_send_platform_message_response(
+    FlEngine* engine,
+    FlEnginePlatformMessageHandle* handle,
+    GBytes* response,
+    GError** error);
```

Comment:
Why does sending a response have an `error` but sending a message doesn't? It seems like an odd asymmetry.

Reply:
`fl_engine_send_platform_message` is an asynchronous call, so the response is in the callback. This includes reporting of errors. This is the GLib convention. Note I have GCancellable objects there too to match the convention even though we don't make use of them. There are some APIs out there that didn't do this and they're pretty hard to retrofit...